README.md
![](https://github.com/allsome/LSYevernote/blob/master/evernote/Assets.xcassets/logo.imageset/logo.png)
[![Build Status](https://travis-ci.org/allsome/LSYEvernote.svg?branch=master)](https://travis-ci.org/allsome/LSYEvernote)
[![](https://img.shields.io/badge/platform-iOS-style--flat--red.svg)]()
[![](https://img.shields.io/badge/Swift-3.0-orange.svg)](https://swift.org)
[![](https://img.shields.io/dub/l/vibe-d.svg)](https://github.com/allsome/LSYevernote/blob/master/LICENSE)

![](http://upload-images.jianshu.io/upload_images/961540-b5cd583b9f1ae6b4.gif?imageMogr2/auto-orient/strip)

## Intro

This demo imitate two impressive effects of [Evernote](https://www.evernote.com) iOS version by using **UICollectionView** and **Controller Transition**, the implementation is not so complicated as it looks. And it's common to make a [CocoaPods](https://guides.cocoapods.org/making/index.html) spec, but unlike so many popular specs, I think their right usage scenarios are limited. It could not be better if you find the project is helpful. By the way, every **Star** or **Fork** means a lot to me.

### CHANGELOG
*  Clone repo refer to CHANGELOG

### Learn more
* Read the blog in [Medium](https://medium.com/@allsome/interaction-in-evernote-5d3b08b2c359#.qyondcdo7)
* 龙的传人点[这里](http://www.jianshu.com/p/dfb857763942)

### Let me know
Please contact oraclelsy@163.com if you have any questions or suggestion about the demo.


 
 


CHANGELOG.md
#### 2.x Releases
- Swift 3.0

#### 1.x Releases
- Swift 2.2


evernote/CollectionViewCell.swift
//
//  CollectionViewCell.swift
//  evernote
//
//  Created by 梁树元 on 10/12/15.
//  Copyright © 2015 com. All rights reserved.
//

import UIKit

class CollectionViewCell: UICollectionViewCell {

    @IBOutlet internal weak var backButton: UIButton!
    @IBOutlet internal weak var titleLine: UIView!
    @IBOutlet internal weak var textView: UITextView! 
    @IBOutlet internal weak var titleLabel: UILabel!
    @IBOutlet internal weak var labelLeadCons: NSLayoutConstraint!
    internal var horizonallyCons = NSLayoutConstraint()
    
    override func awakeFromNib() {
        super.awakeFromNib()
        self.layer.masksToBounds = true
        self.layer.cornerRadius = 5.0
        self.horizonallyCons = NSLayoutConstraint(item: titleLabel, attribute: NSLayoutAttribute.centerX, relatedBy: NSLayoutRelation.equal, toItem: self, attribute: NSLayoutAttribute.centerX, multiplier: 1.0, constant: 0.0)
    }

}


evernote/EvernoteTransition.swift
//
//  EvernoteTransition.swift
//  evernote
//
//  Created by 梁树元 on 10/31/15.
//  Copyright © 2015 com. All rights reserved.
//

import UIKit

class EvernoteTransition: NSObject,UIViewControllerAnimatedTransitioning,UIViewControllerTransitioningDelegate,NoteViewControllerDelegate {
    
    internal var isPresent = true
    var selectCell:CollectionViewCell = CollectionViewCell()
    var visibleCells = [CollectionViewCell]()
    var originFrame:CGRect = CGRect.zero
    var finalFrame:CGRect = CGRect.zero
    var panViewController = UIViewController()
    var listViewController = UIViewController()
    var interactionController = UIPercentDrivenInteractiveTransition()

    func EvernoteTransitionWith(selectCell:CollectionViewCell, visibleCells:[CollectionViewCell], originFrame:CGRect, finalFrame:CGRect, panViewController:UIViewController, listViewController:UIViewController) {
        self.selectCell = selectCell
        self.visibleCells = visibleCells
        self.originFrame = originFrame
        self.finalFrame = finalFrame
        self.panViewController = panViewController
        self.listViewController = listViewController
        let pan = UIScreenEdgePanGestureRecognizer(target: self, action: #selector(EvernoteTransition.handlePanGesture(_:)))
        pan.edges = UIRectEdge.left
        self.panViewController.view.addGestureRecognizer(pan)
    }
    
    // UIViewControllerAnimatedTransitioning
    func transitionDuration(using transitionContext: UIViewControllerContextTransitioning?) -> TimeInterval {
        return 0.45
    }
    
    func animateTransition(using transitionContext: UIViewControllerContextTransitioning) {
        let nextVC = transitionContext.viewController(forKey: UITransitionContextToViewControllerKey)
        transitionContext.containerView.backgroundColor = BGColor
        selectCell.frame = isPresent ? originFrame : finalFrame
        let addView = nextVC?.view
        addView!.isHidden = isPresent ? true : false
        transitionContext.containerView.addSubview(addView!)
        let removeCons = isPresent ? selectCell.labelLeadCons : selectCell.horizonallyCons
        let addCons = isPresent ? selectCell.horizonallyCons : selectCell.labelLeadCons
        selectCell.removeConstraint(removeCons!)
        selectCell.addConstraint(addCons!)
        UIView.animate(withDuration: self.transitionDuration(using: transitionContext), delay: 0.0, options: UIViewAnimationOptions(), animations: { () -> Void in
            for visibleCell:CollectionViewCell in self.visibleCells {
                if visibleCell != self.selectCell {
                    var frame = visibleCell.frame
                    if visibleCell.tag < self.selectCell.tag {
                        let yDistance = self.originFrame.origin.y - self.finalFrame.origin.y + 30
                        let yUpdate = self.isPresent ? yDistance : -yDistance
                        frame.origin.y -= yUpdate
                    }else if visibleCell.tag > self.selectCell.tag{
                        let yDistance = self.finalFrame.maxY - self.originFrame.maxY + 30
                        let yUpdate = self.isPresent ? yDistance : -yDistance
                        frame.origin.y += yUpdate
                    }
                    visibleCell.frame = frame
                    visibleCell.transform = self.isPresent ? CGAffineTransform(scaleX: 0.8, y: 1.0):CGAffineTransform.identity
                }
            }
            self.selectCell.backButton.alpha = self.isPresent ? 1.0 : 0.0
            self.selectCell.titleLine.alpha = self.isPresent ? 1.0 : 0.0
            self.selectCell.textView.alpha = self.isPresent ? 1.0 : 0.0
            self.selectCell.frame = self.isPresent ? self.finalFrame : self.originFrame
            self.selectCell.layoutIfNeeded()
            }) { (stop) -> Void in
                addView!.isHidden = false
                transitionContext.completeTransition(true)
        }

    }
    
    // UIViewControllerTransitioningDelegate
    func animationController(forPresented presented: UIViewController, presenting: UIViewController, source: UIViewController) -> UIViewControllerAnimatedTransitioning? {
        isPresent = true
        selectCell.textView.isScrollEnabled = false
        return self
    }
    
    func animationController(forDismissed dismissed: UIViewController) -> UIViewControllerAnimatedTransitioning? {
        self.isPresent = false
        return self
    }
    
    func interactionControllerForDismissal(using animator: UIViewControllerAnimatedTransitioning) -> UIViewControllerInteractiveTransitioning? {
        return interactionController
    }
    
    func handlePanGesture(_ recognizer:UIScreenEdgePanGestureRecognizer) {
        let view = panViewController.view
        if recognizer.state == UIGestureRecognizerState.began {
            panViewController.dismiss(animated: true, completion: { () -> Void in
            })
        } else if recognizer.state == UIGestureRecognizerState.changed {
            let translation = recognizer.translation(in: view)
            let d = fabs(translation.x / (view?.bounds.width)!)
            interactionController.update(d)
        } else if recognizer.state == UIGestureRecognizerState.ended {
            if recognizer.velocity(in: view).x > 0 {
                finishInteractive()
            } else {
                interactionController.cancel()
                listViewController.present(panViewController, animated: false, completion: { () -> Void in
                    
                })
            }
            interactionController = UIPercentDrivenInteractiveTransition()
        }
    }
    
    // NoteViewControllerDelegate
    func didClickGoBack() {
        panViewController.dismiss(animated: true, completion: { () -> Void in
        })
        finishInteractive()
    }
    
    func finishInteractive() {
        interactionController.finish()
        selectCell.textView.isScrollEnabled = true
    }
    
}


evernote/NoteViewController.swift
//
//  NoteViewController.swift
//  evernote
//
//  Created by 梁树元 on 10/31/15.
//  Copyright © 2015 com. All rights reserved.
//

import UIKit

public protocol NoteViewControllerDelegate:NSObjectProtocol {
    func didClickGoBack()
}

class NoteViewController: UIViewController {
    
    internal weak var delegate: NoteViewControllerDelegate?
    internal var domainColor:UIColor = UIColor()
    internal var titleName:String?
    @IBOutlet weak var totalView: UIView!
    @IBOutlet weak var textView: UITextView!
    @IBOutlet weak var titleLabel: UILabel!

    override func viewDidLoad() {
        super.viewDidLoad()
        totalView.layer.masksToBounds = true
        totalView.layer.cornerRadius = 5.0
        totalView.backgroundColor = domainColor
        titleLabel.text = titleName
        textView.contentOffset = CGPoint(x: 0, y: 0)
        let tap = UITapGestureRecognizer(target: self, action: #selector(NoteViewController.handleTapGesture(_:)))
        self.totalView.addGestureRecognizer(tap)
    }
    
    func handleTapGesture(_ gesture: UITapGestureRecognizer) {
        self.view.endEditing(true)
    }
    
    @IBAction func goBack(_ sender: AnyObject) {
        delegate?.didClickGoBack()
    }
    
    override var preferredStatusBarStyle: UIStatusBarStyle {
        return .lightContent
    }

}


evernoteTests/evernoteTests.swift
//
//  evernoteTests.swift
//  evernoteTests
//
//  Created by 梁树元 on 10/12/15.
//  Copyright © 2015 com. All rights reserved.
//

import XCTest
@testable import evernote

class evernoteTests: XCTestCase {
    
    override func setUp() {
        super.setUp()
        // Put setup code here. This method is called before the invocation of each test method in the class.
    }
    
    override func tearDown() {
        // Put teardown code here. This method is called after the invocation of each test method in the class.
        super.tearDown()
    }
    
    func testExample() {
        // This is an example of a functional test case.
        // Use XCTAssert and related functions to verify your tests produce the correct results.
    }
    
    func testPerformanceExample() {
        // This is an example of a performance test case.
        self.measure {
            // Put the code you want to measure the time of here.
        }
    }
    
}


evernote/CollectionViewLayout.swift
//
//  CollectionViewLayout.swift
//  evernote
//
//  Created by 梁树元 on 10/30/15.
//  Copyright © 2015 com. All rights reserved.
//

import UIKit

public let horizonallyPadding:CGFloat = 10
public let verticallyPadding:CGFloat = 10

public let screenWidth = UIScreen.main.bounds.size.width
public let screenHeight = UIScreen.main.bounds.size.height

public let cellWidth = screenWidth - 2 * horizonallyPadding
private let cellHeight:CGFloat = 45

private let SpringFactor:CGFloat = 10.0

class CollectionViewLayout: UICollectionViewFlowLayout {
    
    override init() {
        super.init()
        itemSize = CGSize(width: cellWidth, height: cellHeight)
        headerReferenceSize = CGSize(width: screenWidth, height: verticallyPadding)
    }
    
    required init(coder aDecoder: NSCoder) {
        super.init(coder: aDecoder)!
        itemSize = CGSize(width: cellWidth, height: cellHeight)
        headerReferenceSize = CGSize(width: screenWidth, height: verticallyPadding)
    }
    
    override func shouldInvalidateLayout(forBoundsChange newBounds: CGRect) -> Bool {
        return true
    }
    
    override func layoutAttributesForElements(in rect: CGRect) -> [UICollectionViewLayoutAttributes]? {
        let offsetY = self.collectionView!.contentOffset.y
        let attrsArray = super.layoutAttributesForElements(in: rect)
        let collectionViewFrameHeight = self.collectionView!.frame.size.height;
        let collectionViewContentHeight = self.collectionView!.contentSize.height;
        let ScrollViewContentInsetBottom = self.collectionView!.contentInset.bottom;
        let bottomOffset = offsetY + collectionViewFrameHeight - collectionViewContentHeight - ScrollViewContentInsetBottom
        let numOfItems = self.collectionView!.numberOfSections
        
        for attr:UICollectionViewLayoutAttributes in attrsArray! {
            if (attr.representedElementCategory == UICollectionElementCategory.cell) {
                var cellRect = attr.frame;
                if offsetY <= 0 {
                    let distance = fabs(offsetY) / SpringFactor;
                    cellRect.origin.y += offsetY + distance * CGFloat(attr.indexPath.section + 1);
                }else if bottomOffset > 0 {
                    let distance = bottomOffset / SpringFactor;
                    cellRect.origin.y += bottomOffset - distance *
                        CGFloat(numOfItems - attr.indexPath.section)
                }
                attr.frame = cellRect;
            }
        }
        return attrsArray;
    }
}


evernoteUITests/evernoteUITests.swift
//
//  evernoteUITests.swift
//  evernoteUITests
//
//  Created by 梁树元 on 10/12/15.
//  Copyright © 2015 com. All rights reserved.
//

import XCTest

class evernoteUITests: XCTestCase {
        
    override func setUp() {
        super.setUp()
        
        // Put setup code here. This method is called before the invocation of each test method in the class.
        
        // In UI tests it is usually best to stop immediately when a failure occurs.
        continueAfterFailure = false
        // UI tests must launch the application that they test. Doing this in setup will make sure it happens for each test method.
        XCUIApplication().launch()

        // In UI tests it’s important to set the initial state - such as interface orientation - required for your tests before they run. The setUp method is a good place to do this.
    }
    
    override func tearDown() {
        // Put teardown code here. This method is called after the invocation of each test method in the class.
        super.tearDown()
    }
    
    func testExample() {
        // Use recording to get started writing UI tests.
        // Use XCTAssert and related functions to verify your tests produce the correct results.
    }
    
}


evernote/CollectionViewController.swift
//
//  CollectionViewController.swift
//  evernote
//
//  Created by 梁树元 on 10/12/15.
//  Copyright © 2015 com. All rights reserved.
//

import UIKit

private let reuseIdentifier = "Cell"
private let topPadding:CGFloat = 20
public let BGColor = UIColor(red: 56.0/255.0, green: 51/255.0, blue: 76/255.0, alpha: 1.0)

class CollectionViewController: UIViewController,UICollectionViewDelegate, UICollectionViewDataSource {
    
    private let colorArray = NSMutableArray()
    private let rowNumber = 15
    private let customTransition = EvernoteTransition()
    private let collectionView = UICollectionView(frame: CGRect(x: 0, y: topPadding, width: screenWidth, height: screenHeight - topPadding), collectionViewLayout: CollectionViewLayout())
    
    override func viewDidLoad() {
        super.viewDidLoad()
        view.backgroundColor = BGColor
        collectionView.backgroundColor = BGColor
        collectionView.dataSource = self
        collectionView.delegate = self
        collectionView.alwaysBounceVertical = false
        collectionView.contentInset = UIEdgeInsetsMake(0, 0, verticallyPadding, 0);

        self.view.addSubview(collectionView)
        let nib = UINib(nibName: "CollectionViewCell", bundle: nil)
        collectionView.register(nib, forCellWithReuseIdentifier: reuseIdentifier)
        
        let random = arc4random() % 360 // 160 arc4random() % 360
        for index in 0 ..< rowNumber {
            let color = UIColor(hue: CGFloat((Int(random) + index * 6)).truncatingRemainder(dividingBy: 360.0) / 360.0, saturation: 0.8, brightness: 1.0, alpha: 1.0)
            colorArray.add(color)
        }
        
    }
    
    override var preferredStatusBarStyle: UIStatusBarStyle {
        return .lightContent
    }
    
    func numberOfSections(in collectionView: UICollectionView) -> Int {
        return colorArray.count
    }
    
     func collectionView(_ collectionView: UICollectionView, numberOfItemsInSection section: Int) -> Int {
        return 1
    }

     func collectionView(_ collectionView: UICollectionView, cellForItemAt indexPath: IndexPath) -> UICollectionViewCell {
        let cell = collectionView.dequeueReusableCell(withReuseIdentifier: reuseIdentifier, for: indexPath) as! CollectionViewCell
        cell.backgroundColor = colorArray.object(at: colorArray.count - 1 - (indexPath as NSIndexPath).section) as? UIColor
        cell.titleLabel.text = "Notebook + " + String((indexPath as NSIndexPath).section + 1)
        cell.titleLine.alpha = 0.0
        cell.textView.alpha = 0.0
        cell.backButton.alpha = 0.0
        cell.tag = (indexPath as NSIndexPath).section
        return cell
    }
    
    func collectionView(_ collectionView: UICollectionView, didSelectItemAt indexPath: IndexPath) {
        let cell = collectionView.cellForItem(at: indexPath) as! CollectionViewCell
        let visibleCells = collectionView.visibleCells as! [CollectionViewCell]
        let storyBoard = UIStoryboard(name: "evernote", bundle: nil)
        let viewController = storyBoard.instantiateViewController(withIdentifier: "Note") as! NoteViewController
        viewController.titleName = cell.titleLabel.text!
        viewController.domainColor = cell.backgroundColor!

        let finalFrame = CGRect(x: 10, y: collectionView.contentOffset.y + 10, width: screenWidth - 20, height: screenHeight - 40)
        self.customTransition.EvernoteTransitionWith(selectCell: cell, visibleCells: visibleCells, originFrame: cell.frame, finalFrame: finalFrame, panViewController:viewController, listViewController: self)
        viewController.transitioningDelegate = self.customTransition
        viewController.delegate = self.customTransition
        self.present(viewController, animated: true) { () -> Void in
        }
    }

}
